### PR TITLE
Add io.github.arunsivaramanneo.GPUViewer

### PR DESCRIPTION
**Pull request by gasinvein**
[GPU-Viewer](/arunsivaramanneo/GPU-Viewer) is an application for viewing varioous GPU information via command-line tools such as glxinfo/vulkaninfo.
I think it can be especially useful for Flatpak users, e.g. for diagnosing driver issues.

### CODE REVIEW

File: io.github.arunsivaramanneo.GPUViewer.appdata.xml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+<component type="desktop-application">
+    <id>io.github.arunsivaramanneo.GPUViewer.desktop</id>
+    <name>GPU-Viewer</name>
+    <summary>A front-end to glxinfo, vulkaninfo, clinfo and es2_info</summary>
+    <description>
+        <p>
+            This project aims to capture all the important details of glxinfo, vulkaninfo and clinfo in a GUI. The project is being developed using python 3 pygobject with GTK3. All the important details are extracted using glxinfo/vulkaninfo/clinfo with the combination of grep, CAT , AWK commands and displayed in the front-end. There is no hard OpenGL Programming involved, until glxinfo, vulkaninfo and clinfo works the GPU-viewer will also work.
+        </p>
+    </description>
+    <categories>
```
Comment (TingPing) [inline io.github.arunsivaramanneo.GPUViewer.appdata.xml]: These should go in the desktop file only.

File: io.github.arunsivaramanneo.GPUViewer.appdata.xml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        <category>System</category>
+        <category>Monitor</category>
+    </categories>
+    <metadata_license>CC0-1.0</metadata_license>
+    <project_license>GPL-3.0</project_license>
+    <url type="homepage">https://github.com/arunsivaramanneo/GPU-Viewer</url>
+    <url type="bugtracker">https://github.com/arunsivaramanneo/GPU-Viewer/issues</url>
+    <url type="donation">https://www.paypal.me/ArunSivaraman</url>
+    <launchable type="desktop-id">net.sourceforge.SimulIDE.desktop</launchable>
```
Comment (TingPing) [inline io.github.arunsivaramanneo.GPUViewer.appdata.xml]: Left-over

File: io.github.arunsivaramanneo.GPUViewer.yml [hunk 3]
```diff
@@ -0,0 +1,119 @@
+id: io.github.arunsivaramanneo.GPUViewer
+runtime: org.gnome.Platform
+sdk: org.gnome.Sdk
+runtime-version: 3.30
```
Comment (TingPing) [inline io.github.arunsivaramanneo.GPUViewer.yml]: Put in quotes `'3.30'`

File: io.github.arunsivaramanneo.GPUViewer.yml [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+            sha256: c173154bbd0d5fb53d732471984def42fb1b14ac85fcb834138fb9518b3e0bef
+        modules:
+          - shared-modules/glew/glew.json
+          - shared-modules/glu/glu-9.0.0.json
+
+          - name: freeglut
+            buildsystem: cmake-ninja
+            config-opts:
+              - -DCMAKE_INSTALL_LIBDIR=/app/lib
+              - -DOpenGL_GL_PREFERENCE=LEGACY
```
Comment (TingPing) [inline io.github.arunsivaramanneo.GPUViewer.yml]: Why legacy?

File: io.github.arunsivaramanneo.GPUViewer.yml [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+            sources:
+              - type: archive
+                url: "https://datapacket.dl.sourceforge.net/project/freeglut/freeglut/3.0.0/freeglut-3.0.0.tar.gz"
+                sha256: 2a43be8515b01ea82bcfa17d29ae0d40bd128342f0930cd1f375f1ff999f76a2
+
+      - name: clinfo
+        buildsystem: simple
+        build-commands:
+          - make -j $FLATPAK_BUILDER_N_JOBS
```
Comment (TingPing) [inline io.github.arunsivaramanneo.GPUViewer.yml]: You can just use `no-autogen: true` instead of calling make manually.

**Comment by TingPing** [comment]
@arunsivaramanneo Hi do you have any interest in having this become an official package? Flatpak allows you to easily distribute the latest version to users on many distros.

File: io.github.arunsivaramanneo.GPUViewer.appdata.xml [hunk 6]
```diff
@@ -0,0 +1,38 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop-application">
+    <id>io.github.arunsivaramanneo.GPUViewer.desktop</id>
```
Comment (TingPing) [inline io.github.arunsivaramanneo.GPUViewer.appdata.xml]: Drop the `.desktop`.

File: io.github.arunsivaramanneo.GPUViewer.yml [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+modules:
+  - name: gpu-viewer
+    buildsystem: simple
+    build-commands:
+      - mkdir -p /app/{bin,share}
+      - cp -a . /app/share/gpu-viewer
+      - install -Dm755 gpu-viewer.sh /app/bin/gpu-viewer
+      - install -Dm644 Images/GPU_Viewer.png /app/share/icons/hicolor/512x512/apps/${FLATPAK_ID}.png
+      - install -Dm644 gpu-viewer.desktop /app/share/applications/${FLATPAK_ID}.desktop
+      - desktop-file-edit --remove-key=Terminal  --remove-key=version /app/share/applications/${FLATPAK_ID}.desktop
```
Comment (bilelmoussaoui) [inline io.github.arunsivaramanneo.GPUViewer.yml]: Those could be pushed to upstream.

File: io.github.arunsivaramanneo.GPUViewer.yml [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+      - desktop-file-edit --set-key=Terminal --set-value=false /app/share/applications/${FLATPAK_ID}.desktop
+      - desktop-file-edit --set-key=Exec --set-value=gpu-viewer /app/share/applications/${FLATPAK_ID}.desktop
+      - desktop-file-edit --set-key=Icon --set-value=${FLATPAK_ID} /app/share/applications/${FLATPAK_ID}.desktop
+      - install -Dm644 -t /app/share/appdata ${FLATPAK_ID}.appdata.xml
+    sources:
+      - type: archive
+        url: "https://github.com/arunsivaramanneo/GPU-Viewer/archive/v1.14.tar.gz"
+        sha256: e81804dca0576cb7d3ccda78f3b34aeb3b16c99110a41b60b0c661737dbc609b
+      - type: script
```
Comment (bilelmoussaoui) [inline io.github.arunsivaramanneo.GPUViewer.yml]: Why is this needed? You could just link the GPUViewer.py to /app/bin  & use it in `command` property (will need to update the desktop file too)

File: io.github.arunsivaramanneo.GPUViewer.yml [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+              - type: archive
+                url: "https://github.com/KhronosGroup/Vulkan-Headers/archive/sdk-1.1.82.0.tar.gz"
+                sha256: df73da07d547cfbe88a797802401ea8225e4844e13d4fde52a7cb6e00e5179e5
+
+      - name: mesa-demos
+        config-opts:
+          - --bindir=/app/lib/mesa-demos
+        post-install:
+          - mv -v /app/lib/mesa-demos/*info /app/bin/
+          - rm -r /app/lib/mesa-demos
```
Comment (bilelmoussaoui) [inline io.github.arunsivaramanneo.GPUViewer.yml]: Should be put in cleanup instead of doing that here.

File: io.github.arunsivaramanneo.GPUViewer.yml [hunk 10]
```diff
@@ -0,0 +1,119 @@
+id: io.github.arunsivaramanneo.GPUViewer
+runtime: org.gnome.Platform
```
Comment (bilelmoussaoui) [inline io.github.arunsivaramanneo.GPUViewer.yml]: Is there any specific reasons to use the gnome runtime? as far as I can tell, the application uses Gtk only from GNOME. You should be using freedestkop runtime instead which has Gtk.

File: io.github.arunsivaramanneo.GPUViewer.appdata.xml [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+        <screenshot>
+            <image>https://user-images.githubusercontent.com/30646692/42731015-174af686-8822-11e8-8dfb-5baad1cface1.png</image>
+        </screenshot>
+        <screenshot>
+            <image>https://user-images.githubusercontent.com/30646692/42731016-18ca0628-8822-11e8-90c1-def21ee5d0bb.png</image>
+        </screenshot>
+    </screenshots>
+    <releases>
+        <release version="1.14" date="2018-09-30" />
+    </releases>
```
Comment (bilelmoussaoui) [inline io.github.arunsivaramanneo.GPUViewer.appdata.xml]: OARS tags are missing in the appdata file.

File: io.github.arunsivaramanneo.GPUViewer.yml [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+      - name: clinfo
+        buildsystem: simple
+        build-commands:
+          - make -j $FLATPAK_BUILDER_N_JOBS
+          - install -Dm755 -t /app/bin/ clinfo
+        sources:
+          - type: archive
+            url: "https://github.com/Oblomov/clinfo/archive/2.2.18.04.06.tar.gz"
+            sha256: f77021a57b3afcdebc73107e2254b95780026a9df9aa4f8db6aff11c03f0ec6c
+        modules:
```
Comment (bilelmoussaoui) [inline io.github.arunsivaramanneo.GPUViewer.yml]: Why did you repeat modules here? I don't see why you are using multiple times the modules keyword.

**Comment by arunsivaramanneo** [comment]
> @arunsivaramanneo Hi do you have any interest in having this become an official package? Flatpak allows you to easily distribute the latest version to users on many distros.



> @arunsivaramanneo Hi do you have any interest in having this become an official package? Flatpak allows you to easily distribute the latest version to users on many distros.

@gasinvein Thanks for doing this, I have been trying to get this done long back and haven't got a chance to look into it. @TingPing Yes, I am trying to make this as an Official Package. Do let me know if anything i need to do

File: io.github.arunsivaramanneo.GPUViewer.yml (hunk 4 above)
Comment (gasinvein) [reply io.github.arunsivaramanneo.GPUViewer.yml]: It fails to build with GLVND. Help appreciated.

File: io.github.arunsivaramanneo.GPUViewer.yml (hunk 5 above)
Comment (gasinvein) [reply io.github.arunsivaramanneo.GPUViewer.yml]: `make install` is broken here, so I would need to add `no-make-install: true` and run `install` manually anyway. Using `buildsystem: simple` just makes it a bit simplier.

File: io.github.arunsivaramanneo.GPUViewer.yml (hunk 7 above)
Comment (gasinvein) [reply io.github.arunsivaramanneo.GPUViewer.yml]: I'll send PR later.

File: io.github.arunsivaramanneo.GPUViewer.yml (hunk 8 above)
Comment (gasinvein) [reply io.github.arunsivaramanneo.GPUViewer.yml]: No, I can't. The app hardcodes some relative paths, e.g. [here](/arunsivaramanneo/GPU-Viewer/blob/64f1e671c09350e3128eb07c04cdec2f5421d811/Files/Const.py).

File: io.github.arunsivaramanneo.GPUViewer.yml (hunk 10 above)
Comment (gasinvein) [reply io.github.arunsivaramanneo.GPUViewer.yml]: FDO runtime is missing python-gobject.

File: io.github.arunsivaramanneo.GPUViewer.yml (hunk 12 above)
Comment (gasinvein) [reply io.github.arunsivaramanneo.GPUViewer.yml]: I use nested modules because it represents modules dependencies, and makes reading the manifest easier than just flat list of modules.

**Comment by nedrichards** [comment]
bot, build io.github.arunsivaramanneo.GPUViewer

File: io.github.arunsivaramanneo.GPUViewer.appdata.xml [hunk 13]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop-application">
+    <id>io.github.arunsivaramanneo.GPUViewer</id>
```
Comment (TingPing) [inline io.github.arunsivaramanneo.GPUViewer.appdata.xml]: Missing `<launchable>`

**Comment by TingPing** [comment]
> @TingPing Yes, I am trying to make this as an Official Package. Do let me know if anything i need to do

That is great to hear. Basically:

- Add this appstream file upstream
- Rename your icons and desktop file to match app-id
- Add an actual build system so it can be reasonably manage (i recommend https://meson.build personally)

File: io.github.arunsivaramanneo.GPUViewer.yml (hunk 10 above)
Comment (gasinvein) [reply io.github.arunsivaramanneo.GPUViewer.yml]: @TingPing should I switch to FDO runtime and and bundle everything needed, or Gnome runtime is fine?

**Comment by nedrichards** [comment]
bot, build io.github.arunsivaramanneo.GPUViewer

File: io.github.arunsivaramanneo.GPUViewer.appdata.xml [hunk 14]
```diff
@@ -1,6 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <component type="desktop-application">
     <id>io.github.arunsivaramanneo.GPUViewer</id>
+    <launchable type="desktop-id">net.sourceforge.SimulIDE.desktop</launchable>
```
Comment (TingPing) [inline io.github.arunsivaramanneo.GPUViewer.appdata.xml]: ...

File: io.github.arunsivaramanneo.GPUViewer.yml [hunk 15]
```diff
@@ -3,7 +3,7 @@ runtime: org.freedesktop.Platform
 sdk: org.freedesktop.Sdk
 runtime-version: "18.08"
 finish-args:
-  - --device=dri
+  - --device=all # OpenCL requires /dev/nvidia-uvm
```
Comment (TingPing) [inline io.github.arunsivaramanneo.GPUViewer.yml]: This sounds like a bug to report to flatpak/flatpak.

Comment (gasinvein) [reply io.github.arunsivaramanneo.GPUViewer.yml]: OpenCL on Nvidia in FDO runtime 18.08 seems broken, but can be fixed locally.

Comment (TingPing) [reply io.github.arunsivaramanneo.GPUViewer.yml]: > In the runtime files, /etc/OpenCL should point to ../lib/x86_64-linux-gnu/GL/OpenCL, and not to ../lib/GL/OpenCL
> In runtime's metadata, OpenCL/vendors should be added to merge-dirs

Please report that here: https://gitlab.com/freedesktop-sdk/freedesktop-sdk

> The .GL.nvidia-* extension lacks OpenCL symlink to extra/OpenCL

Please report that here: https://github.com/flatpak/freedesktop-sdk-images

**Comment by nedrichards** [comment]
bot, build io.github.arunsivaramanneo.GPUViewer

**Comment by barthalion** [comment]
With the issue fixed in FDo SDK (at least in 18.08), are there any blockers?

**Comment by nedrichards** [comment]
Repository has been created: https://github.com/flathub/io.github.arunsivaramanneo.GPUViewer

**Comment by nedrichards** [comment]
Er, a blocker is that the build doesn't work :-( https://flathub.org/builds/#/builders/2/builds/5889 which is unfortunate given that I just pushed it - the app won't be published until it does, so there's no impact to users. I'll file an issue on the project and we can work on it there.

**Comment by gasinvein** [comment]
@nedrichards  Please don't forget to add the actual app developer @arunsivaramanneo as a collaborator.

**Comment by arunsivaramanneo** [comment]
Thanks team, i did  install it on my Linux Mint machine having switchable Graphics through Flatpak, I can see that the application works fine on Intel GPU as default but when i switch to nvidia prime i don't see both the gpu listed on Vulkan tab only the first gpu is displayed and not sure why OpenGL information is missing in OpenGL tab. This only happen when i switch to nvidia as default. should work fine on single gpu system

screenshot from 2018-11-07 08-36-42

**Comment by TingPing** [comment]
There needs to be work done for prime to work with Flatpak.

**Comment by winehub** [comment]
@TingPing 

I did my own one from "unkown". It is shrinked to 4MB. You can take a view on that at:

https://github.com/winehub/GPU-Viewer/blob/master/online.winehub.GPUViewer.yml
